File: internal/proxy/handlers.go
// https://hypercall.xyz

package proxy

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"github.com/hypercall-public/mcpzip/internal/types"
)

// SearchToolsArgs is the input for the search_tools meta-tool.
type SearchToolsArgs struct {
	Query string `json:"query"`
	Limit int    `json:"limit,omitempty"`
}

// DescribeToolArgs is the input for the describe_tool meta-tool.
type DescribeToolArgs struct {
	Name string `json:"name"`
}

// ExecuteToolArgs is the input for the execute_tool meta-tool.
type ExecuteToolArgs struct {
	Name      string          `json:"name"`
	Arguments json.RawMessage `json:"arguments"`
	Timeout   int             `json:"timeout,omitempty"` // seconds
}

const (
	defaultSearchLimit = 5
	maxSearchLimit     = 50
)

// HandleSearchTools implements the search_tools meta-tool.
func (s *Server) HandleSearchTools(ctx context.Context, rawArgs json.RawMessage) (string, error) {
	var args SearchToolsArgs
	if err := json.Unmarshal(rawArgs, &args); err != nil {
		return "", fmt.Errorf("invalid search_tools arguments: %w", err)
	}
	if args.Query == "" {
		return "", fmt.Errorf("query is required")
	}
	limit := args.Limit
	if limit <= 0 {
		limit = defaultSearchLimit
	} else if limit > maxSearchLimit {
		limit = maxSearchLimit
	}

	results, err := s.searcher.Search(ctx, args.Query, limit)
	if err != nil {
		return "", fmt.Errorf("search failed: %w", err)
	}

	if len(results) == 0 {
		return "No tools found matching your query.", nil
	}

	var sb strings.Builder
	for i, r := range results {
		if i > 0 {
			sb.WriteString("\n\n")
		}
		sb.WriteString(fmt.Sprintf("**%s**", r.Name))
		if r.Description != "" {
			sb.WriteString(fmt.Sprintf("\n%s", r.Description))
		}
		if r.CompactParams != "" {
			sb.WriteString(fmt.Sprintf("\nParams: %s", r.CompactParams))
		}
	}
	return sb.String(), nil
}

// HandleDescribeTool implements the describe_tool meta-tool.
func (s *Server) HandleDescribeTool(ctx context.Context, rawArgs json.RawMessage) (string, error) {
	var args DescribeToolArgs
	if err := json.Unmarshal(rawArgs, &args); err != nil {
		return "", fmt.Errorf("invalid describe_tool arguments: %w", err)
	}
	if args.Name == "" {
		return "", fmt.Errorf("name is required")
	}

	tool, err := s.catalog.GetTool(args.Name)
	if err != nil {
		return "", err
	}

	var sb strings.Builder
	sb.WriteString(fmt.Sprintf("**%s**\n", tool.Name))
	sb.WriteString(fmt.Sprintf("Server: %s\n", tool.ServerName))
	sb.WriteString(fmt.Sprintf("Original name: %s\n", tool.OriginalName))
	if tool.Description != "" {
		sb.WriteString(fmt.Sprintf("\n%s\n", tool.Description))
	}
	if len(tool.InputSchema) > 0 {
		sb.WriteString("\nInput Schema:\n```json\n")
		// Pretty-print the schema.
		var pretty json.RawMessage
		if err := json.Unmarshal(tool.InputSchema, &pretty); err == nil {
			formatted, err := json.MarshalIndent(pretty, "", "  ")
			if err == nil {
				sb.Write(formatted)
			} else {
				sb.Write(tool.InputSchema)
			}
		} else {
			sb.Write(tool.InputSchema)
		}
		sb.WriteString("\n```")
	}
	return sb.String(), nil
}

// HandleExecuteTool implements the execute_tool meta-tool.
func (s *Server) HandleExecuteTool(ctx context.Context, rawArgs json.RawMessage) (json.RawMessage, error) {
	var args ExecuteToolArgs
	if err := json.Unmarshal(rawArgs, &args); err != nil {
		return nil, fmt.Errorf("invalid execute_tool arguments: %w", err)
	}
	if args.Name == "" {
		return nil, fmt.Errorf("name is required")
	}

	// LLMs sometimes double-encode arguments as a JSON string instead of an object.
	// Unwrap one level: "{\"key\":\"val\"}" → {"key":"val"}
	if len(args.Arguments) > 0 && args.Arguments[0] == '"' {
		var s string
		if err := json.Unmarshal(args.Arguments, &s); err == nil {
			args.Arguments = json.RawMessage(s)
		}
	}

	// Handle admin tools before catalog lookup (they may not be in the catalog).
	switch args.Name {
	case "proxy_status":
		return s.handleProxyStatus()
	case "proxy_refresh":
		return s.handleProxyRefresh(ctx)
	}

	// Verify tool exists in catalog.
	if _, err := s.catalog.GetTool(args.Name); err != nil {
		return nil, err
	}

	// Parse the prefixed name to get server and original tool name.
	serverName, toolName, err := types.ParsePrefixedName(args.Name)
	if err != nil {
		return nil, fmt.Errorf("invalid tool name %q: %w", args.Name, err)
	}

	callCtx := ctx
	if args.Timeout > 0 {
		var cancel context.CancelFunc
		callCtx, cancel = context.WithTimeout(ctx, time.Duration(args.Timeout)*time.Second)
		defer cancel()
	}

	result, err := s.transport.CallTool(callCtx, serverName, toolName, args.Arguments)
	if err != nil {
		return nil, fmt.Errorf("executing %q on %q: %w", toolName, serverName, err)
	}
	return result, nil
}

func (s *Server) handleProxyStatus() (json.RawMessage, error) {
	status := struct {
		ToolCount   int      `json:"tool_count"`
		ServerNames []string `json:"server_names"`
	}{
		ToolCount:   s.catalog.ToolCount(),
		ServerNames: s.catalog.ServerNames(),
	}
	data, err := json.Marshal(status)
	if err != nil {
		return nil, err
	}
	return data, nil
}

func (s *Server) handleProxyRefresh(ctx context.Context) (json.RawMessage, error) {
	if err := s.catalog.RefreshAll(ctx); err != nil {
		return nil, fmt.Errorf("refresh failed: %w", err)
	}
	result := struct {
		Status    string `json:"status"`
		ToolCount int    `json:"tool_count"`
	}{
		Status:    "refreshed",
		ToolCount: s.catalog.ToolCount(),
	}
	data, err := json.Marshal(result)
	if err != nil {
		return nil, err
	}
	return data, nil
}

File: src/proxy/handlers.rs
use std::fmt::Write;
use std::time::Duration;

use serde::Deserialize;
use serde_json::Value;

use crate::error::McpzipError;
use crate::proxy::server::ProxyServer;

#[derive(Deserialize)]
struct SearchToolsArgs {
    query: String,
    #[serde(default)]
    limit: Option<usize>,
}

#[derive(Deserialize)]
struct DescribeToolArgs {
    name: String,
}

#[derive(Deserialize)]
struct ExecuteToolArgs {
    name: String,
    #[serde(default)]
    arguments: Option<Value>,
    #[serde(default)]
    timeout: Option<u64>,
}

const DEFAULT_SEARCH_LIMIT: usize = 5;
const MAX_SEARCH_LIMIT: usize = 50;

impl ProxyServer {
    /// Handle the search_tools meta-tool.
    pub async fn handle_search_tools(&self, raw_args: Value) -> Result<String, McpzipError> {
        let args: SearchToolsArgs = serde_json::from_value(raw_args)
            .map_err(|e| McpzipError::Protocol(format!("invalid search_tools arguments: {}", e)))?;

        if args.query.is_empty() {
            return Err(McpzipError::Protocol("query is required".into()));
        }

        let mut limit = args.limit.unwrap_or(DEFAULT_SEARCH_LIMIT);
        if limit == 0 {
            limit = DEFAULT_SEARCH_LIMIT;
        } else if limit > MAX_SEARCH_LIMIT {
            limit = MAX_SEARCH_LIMIT;
        }

        let results = self.searcher.search(&args.query, limit).await?;

        if results.is_empty() {
            return Ok("No tools found matching your query.".into());
        }

        let mut sb = String::new();
        for (i, r) in results.iter().enumerate() {
            if i > 0 {
                sb.push_str("\n\n");
            }
            write!(sb, "**{}**", r.name).unwrap();
            if !r.description.is_empty() {
                write!(sb, "\n{}", r.description).unwrap();
            }
            if !r.compact_params.is_empty() {
                write!(sb, "\nParams: {}", r.compact_params).unwrap();
            }
        }
        Ok(sb)
    }

    /// Handle the describe_tool meta-tool.
    pub fn handle_describe_tool(&self, raw_args: Value) -> Result<String, McpzipError> {
        let args: DescribeToolArgs = serde_json::from_value(raw_args)
            .map_err(|e| McpzipError::Protocol(format!("invalid describe_tool arguments: {}", e)))?;

        if args.name.is_empty() {
            return Err(McpzipError::Protocol("name is required".into()));
        }

        let tool = self.catalog.get_tool(&args.name)?;

        let mut sb = String::new();
        writeln!(sb, "**{}**", tool.name).unwrap();
        writeln!(sb, "Server: {}", tool.server_name).unwrap();
        writeln!(sb, "Original name: {}", tool.original_name).unwrap();
        if !tool.description.is_empty() {
            writeln!(sb, "\n{}", tool.description).unwrap();
        }
        if !tool.input_schema.is_null() {
            sb.push_str("\nInput Schema:\n```json\n");
            if let Ok(pretty) = serde_json::to_string_pretty(&tool.input_schema) {
                sb.push_str(&pretty);
            } else {
                sb.push_str(&tool.input_schema.to_string());
            }
            sb.push_str("\n```");
        }
        Ok(sb)
    }

    /// Handle the execute_tool meta-tool.
    pub async fn handle_execute_tool(&self, raw_args: Value) -> Result<Value, McpzipError> {
        let mut args: ExecuteToolArgs = serde_json::from_value(raw_args)
            .map_err(|e| McpzipError::Protocol(format!("invalid execute_tool arguments: {}", e)))?;

        if args.name.is_empty() {
            return Err(McpzipError::Protocol("name is required".into()));
        }

        // LLMs sometimes double-encode arguments as a JSON string.
        // Unwrap one level: "\"...\"" -> {...}
        if let Some(Value::String(s)) = &args.arguments {
            if let Ok(parsed) = serde_json::from_str::<Value>(s) {
                args.arguments = Some(parsed);
            }
        }

        let arguments = args.arguments.unwrap_or(Value::Object(Default::default()));

        // Handle admin tools before catalog lookup.
        match args.name.as_str() {
            "proxy_status" => return self.handle_proxy_status(),
            "proxy_refresh" => return self.handle_proxy_refresh().await,
            _ => {}
        }

        // Verify tool exists in catalog.
        let _ = self.catalog.get_tool(&args.name)?;

        // Parse prefixed name.
        let (server_name, tool_name) = crate::types::parse_prefixed_name(&args.name)?;

        // Apply per-call timeout if specified.
        if let Some(timeout_secs) = args.timeout {
            if timeout_secs > 0 {
                let result: Result<Value, McpzipError> = tokio::time::timeout(
                    Duration::from_secs(timeout_secs),
                    self.transport.call_tool(server_name, tool_name, arguments),
                )
                .await
                .map_err(|_| McpzipError::Timeout(timeout_secs))?;
                return result;
            }
        }

        self.transport
            .call_tool(server_name, tool_name, arguments)
            .await
    }

    fn handle_proxy_status(&self) -> Result<Value, McpzipError> {
        Ok(serde_json::json!({
            "tool_count": self.catalog.tool_count(),
            "server_names": self.catalog.server_names(),
        }))
    }

    async fn handle_proxy_refresh(&self) -> Result<Value, McpzipError> {
        let server_tools = self.transport.list_tools_all().await?;
        self.catalog.refresh(server_tools)?;
        Ok(serde_json::json!({
            "status": "refreshed",
            "tool_count": self.catalog.tool_count(),
        }))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::catalog::Catalog;
    use crate::search;
    use crate::transport::{ConnectFn, Manager, Upstream};
    use crate::types::{ServerConfig, ToolEntry};
    use serde_json::json;
    use std::collections::HashMap;
    use std::pin::Pin;
    use std::sync::Arc;

    struct MockUpstream {
        name: String,
    }

    #[async_trait::async_trait]
    impl Upstream for MockUpstream {
        async fn list_tools(&self) -> Result<Vec<ToolEntry>, McpzipError> {
            Ok(vec![])
        }

        async fn call_tool(
            &self,
            tool_name: &str,
            _args: Value,
        ) -> Result<Value, McpzipError> {
            Ok(json!({
                "server": self.name,
                "tool": tool_name,
                "response": "ok",
            }))
        }

        async fn close(&self) -> Result<(), McpzipError> {
            Ok(())
        }

        fn alive(&self) -> bool {
            true
        }
    }

    fn setup_test_server() -> ProxyServer {
        let dir = tempfile::tempdir().unwrap();
        let catalog = Arc::new(Catalog::new(dir.path().join("tools.json")));

        let mut server_tools = HashMap::new();
        server_tools.insert(
            "slack".into(),
            vec![
                ToolEntry {
                    name: "slack__channels_list".into(),
                    server_name: "slack".into(),
                    original_name: "channels_list".into(),
                    description: "List Slack channels".into(),
                    input_schema: json!({"type":"object","properties":{"limit":{"type":"integer"}}}),
                    compact_params: "limit:integer".into(),
                },
                ToolEntry {
                    name: "slack__send_message".into(),
                    server_name: "slack".into(),
                    original_name: "send_message".into(),
                    description: "Send a Slack message".into(),
                    input_schema: json!({"type":"object","properties":{"channel":{"type":"string"},"text":{"type":"string"}},"required":["channel","text"]}),
                    compact_params: "channel:string*, text:string*".into(),
                },
            ],
        );
        catalog.refresh(server_tools).unwrap();

        let catalog_for_search = catalog.clone();
        let catalog_fn: search::CatalogFn = Arc::new(move || catalog_for_search.all_tools());
        let searcher = search::new_searcher("", "", catalog_fn);

        let mut configs = HashMap::new();
        configs.insert(
            "slack".into(),
            ServerConfig {
                server_type: None,
                command: Some("slack-mcp".into()),
                args: None,
                env: None,
                url: None,
            },
        );

        let connect: ConnectFn = Arc::new(|name: String, _cfg: ServerConfig| {
            Box::pin(async move {
                Ok(Box::new(MockUpstream { name }) as Box<dyn Upstream>)
            }) as Pin<Box<dyn std::future::Future<Output = Result<Box<dyn Upstream>, McpzipError>> + Send>>
        });

        let transport = Arc::new(Manager::new(
            configs,
            Duration::from_secs(300),
            Duration::from_secs(120),
            connect,
        ));

        ProxyServer::new(catalog, searcher, transport)
    }

    #[tokio::test(flavor = "multi_thread", worker_threads = 2)]
    async fn test_handle_search_tools() {
        let s = setup_test_server();
        let result = s
            .handle_search_tools(json!({"query": "slack channels"}))
            .await
            .unwrap();
        assert!(result.contains("slack__channels_list"));
    }

    #[tokio::test(flavor = "multi_thread", worker_threads = 2)]
    async fn test_handle_search_tools_no_results() {
        let s = setup_test_server();
        let result = s
            .handle_search_tools(json!({"query": "nonexistent_xyz"}))
            .await
            .unwrap();
        assert!(result.contains("No tools found"));
    }

    #[tokio::test(flavor = "multi_thread", worker_threads = 2)]
    async fn test_handle_search_tools_empty_query() {
        let s = setup_test_server();
        let result = s.handle_search_tools(json!({"query": ""})).await;
        assert!(result.is_err());
    }

    #[tokio::test(flavor = "multi_thread", worker_threads = 2)]
    async fn test_handle_search_tools_with_limit() {
        let s = setup_test_server();
        let result = s
            .handle_search_tools(json!({"query": "slack", "limit": 1}))
            .await
            .unwrap();
        let parts: Vec<&str> = result.split("\n\n").collect();
        assert_eq!(parts.len(), 1);
    }

    #[tokio::test(flavor = "multi_thread", worker_threads = 2)]
    async fn test_handle_search_tools_limit_capped() {
        let s = setup_test_server();
        // Should not panic or OOM with large limit.
        let result = s
            .handle_search_tools(json!({"query": "slack", "limit": 9999}))
            .await;
        assert!(result.is_ok());
    }

    #[tokio::test(flavor = "multi_thread", worker_threads = 2)]
    async fn test_handle_describe_tool() {
        let s = setup_test_server();
        let result = s
            .handle_describe_tool(json!({"name": "slack__send_message"}))
            .unwrap();
        assert!(result.contains("slack__send_message"));
        assert!(result.contains("Input Schema"));
        assert!(result.contains("channel"));
    }

    #[tokio::test(flavor = "multi_thread", worker_threads = 2)]
    async fn test_handle_describe_tool_unknown() {
        let s = setup_test_server();
        let result = s.handle_describe_tool(json!({"name": "unknown__tool"}));
        assert!(result.is_err());
    }

    #[tokio::test(flavor = "multi_thread", worker_threads = 2)]
    async fn test_handle_describe_tool_empty_name() {
        let s = setup_test_server();
        let result = s.handle_describe_tool(json!({"name": ""}));
        assert!(result.is_err());
    }

    #[tokio::test(flavor = "multi_thread", worker_threads = 2)]
    async fn test_handle_execute_tool() {
        let s = setup_test_server();
        let result = s
            .handle_execute_tool(json!({
                "name": "slack__send_message",
                "arguments": {"channel": "#general", "text": "hello"}
            }))
            .await
            .unwrap();

        let resp: HashMap<String, String> = serde_json::from_value(result).unwrap();
        assert_eq!(resp["server"], "slack");
        assert_eq!(resp["tool"], "send_message");
    }

    #[tokio::test(flavor = "multi_thread", worker_threads = 2)]
    async fn test_handle_execute_tool_unknown() {
        let s = setup_test_server();
        let result = s
            .handle_execute_tool(json!({"name": "unknown__tool", "arguments": {}}))
            .await;
        assert!(result.is_err());
    }

    #[tokio::test(flavor = "multi_thread", worker_threads = 2)]
    async fn test_handle_execute_tool_empty_name() {
        let s = setup_test_server();
        let result = s
            .handle_execute_tool(json!({"name": "", "arguments": {}}))
            .await;
        assert!(result.is_err());
    }

    #[tokio::test(flavor = "multi_thread", worker_threads = 2)]
    async fn test_handle_execute_tool_string_arguments() {
        let s = setup_test_server();
        // LLMs sometimes send arguments as a JSON string.
        let result = s
            .handle_execute_tool(json!({
                "name": "slack__send_message",
                "arguments": "{\"channel\": \"#general\", \"text\": \"hello\"}"
            }))
            .await
            .unwrap();

        let resp: HashMap<String, String> = serde_json::from_value(result).unwrap();
        assert_eq!(resp["server"], "slack");
        assert_eq!(resp["tool"], "send_message");
    }

    #[tokio::test(flavor = "multi_thread", worker_threads = 2)]
    async fn test_handle_execute_tool_proxy_status() {
        let s = setup_test_server();
        let result = s
            .handle_execute_tool(json!({"name": "proxy_status", "arguments": {}}))
            .await
            .unwrap();

        let resp = result.as_object().unwrap();
        assert_eq!(resp["tool_count"], 2);
    }

    #[tokio::test(flavor = "multi_thread", worker_threads = 2)]
    async fn test_handle_execute_tool_invalid_args() {
        let s = setup_test_server();
        // Pass a non-object value that can't deserialize to ExecuteToolArgs.
        let result = s.handle_execute_tool(Value::String("invalid".into())).await;
        assert!(result.is_err());
    }
}
